kms-connector/crates/kms-worker/src/core/event_processor/s3.rs

```diff
@@ -211,70 +219,136 @@ where
         let mut sns_ciphertext_materials = Vec::new();
-        for sns_material in sns_materials {
-            let extracted_ct_handle = sns_material.ctHandle.to_vec();
-            let extracted_sns_ciphertext_digest = sns_material.snsCiphertextDigest.to_vec();
-            let coprocessor_addresses = sns_material.coprocessorTxSenderAddresses;
 
+        for sns_material in sns_materials {
             // Get S3 URL and retrieve ciphertext
             // 1. For each SNS material, we try to retrieve its ciphertext from multiple possible S3 URLs
+            //    1.1. We try to fetch the ciphertext for `self.s3_ct_retrieval_retries` times fo each S3 URL
             // 2. Once we successfully retrieve a ciphertext from any of those URLs, we break out of the S3 URLs loop
             // 3. Then we continue processing the next SNS material in the outer loop
             let s3_urls = self
-                .prefetch_coprocessor_buckets(coprocessor_addresses)
+                .prefetch_coprocessor_buckets(sns_material.coprocessorTxSenderAddresses)
                 .await;
 
+            let ct_digest = sns_material.snsCiphertextDigest.as_slice();
+            let ct_digest_hex = hex::encode(ct_digest);
             if s3_urls.is_empty() {
-                warn!(
-                    "No S3 URLs found for ciphertext digest {}",
-                    alloy::hex::encode(&extracted_sns_ciphertext_digest)
-                );
+                warn!("No S3 URLs found for ciphertext digest {ct_digest_hex}",);
                 continue;
             }
 
-            let mut ciphertext_retrieved = false;
-            for s3_url in s3_urls {
-                match retrieve_s3_ciphertext(
-                    s3_url.clone(),
-                    extracted_sns_ciphertext_digest.clone(),
-                )
+            match self
+                .retrieve_s3_ciphertext_with_retry(s3_urls, ct_digest, &ct_digest_hex)
                 .await
+            {
+                Some(ciphertext) => {
+                    sns_ciphertext_materials.push((sns_material.ctHandle.to_vec(), ciphertext));
+                }
+                None => error!(
+                    "Failed to retrieve ciphertext for digest {ct_digest_hex} from any S3 URL"
+                ),
+            }
+        }
+
+        sns_ciphertext_materials
+    }
+
+    /// Retrieves a ciphertext from S3 with `self.s3_ct_retrieval_retries` retries.
+    pub async fn retrieve_s3_ciphertext_with_retry(
+        &self,
+        s3_urls: Vec<String>,
+        ciphertext_digest: &[u8],
+        digest_hex: &str,
+    ) -> Option<Vec<u8>> {
+        for i in 1..=self.s3_ciphertext_retrieval_retries {
+            for s3_url in s3_urls.iter() {
+                match self
+                    .retrieve_s3_ciphertext(s3_url, ciphertext_digest, digest_hex)
+                    .await
                 {
                     Ok(ciphertext) => {
                         info!(
```

Comment:
Let’s consider whether it’s worth retrying a potentially dead URL, or if we’d be better off moving on to other URLs that might be healthy. May be spend minimal time and retries on each as your defaults are a bit time consuming?